[FLINK-27066] Reintroduce e2e tests in ES as Java tests.

afedulov opened the pull request:
This PR introduces Java-based E2E tests for the Elasticsearch connectors based on the Sink testing framework and removes the legacy bash-based tests [TBD].

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/pom.xml:
```diff
@@ -0,0 +1,158 @@
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-connector-test-utils</artifactId>
+			<version>1.16-SNAPSHOT</version>
```
I guess all `1.16-SNAPTHOT` should be `${project.version}`

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/pom.xml:
```diff
@@ -0,0 +1,158 @@
+								<transformer
+									implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+									<mainClass>
+										org.apache.flink.streaming.tests.Elasticsearch6SinkExample
```
Is the `Elasticsearch6SinkExample` used anywhere still? If not we can remove most of the complexity of the pom.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/main/java/com/apache/flink/streaming/tests/TupleC2.java:
```diff
@@ -0,0 +1,30 @@
+import org.apache.flink.api.java.tuple.Tuple2;
+
+/** A {@link Comparable} implementation of {@link Tuple2}. */
+public class TupleC2<T0 extends Comparable<? super T0>, T1 extends Comparable<? super T1>>
```
I would much rather have a dedicated class for the test results than introduce a new `Tuple` class.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6DataReader.java:
```diff
@@ -0,0 +1,85 @@
+                                                    .trackTotalHits(true)),
+                            RequestOptions.DEFAULT);
+            SearchHit[] searchHits = response.getHits().getHits();
+            for (SearchHit searchHit : searchHits) {
+                TupleC2<String, String> hit =
+                        TupleC2.of(
+                                searchHit.getId(),
+                                searchHit.getSourceAsMap().get("data").toString());
+                result.add(hit);
```
Nit: I think you can replace this whole block with java streams API.
```java
from += searchHits.length;
return Arrays.stream(searchHits())
    .map(...)
    .collect(...);
```

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6DataReader.java:
```diff
@@ -0,0 +1,85 @@
+    private int from = 0;
+
+    public Elasticsearch6DataReader(RestHighLevelClient client, String indexName, int pageLength) {
+        this.client = client;
+        this.indexName = indexName;
+        this.pageLength = pageLength;
```
please add `checkNotNull` for the non nullable parameters.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6DataReader.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public void close() throws Exception {
+        if (client != null) {
```
Can the client be null?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkE2ECase.java:
```diff
@@ -0,0 +1,78 @@
+                    .fromContainer(
+                            new ElasticsearchContainer(
+                                            DockerImageName.parse(
+                                                    DockerImageVersions.ELASTICSEARCH_7))
```
```suggestion
                                                    DockerImageVersions.ELASTICSEARCH_6))
```

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkE2ECase.java:
```diff
@@ -0,0 +1,78 @@
+            new Elasticsearch6SinkExternalContextFactory(
+                    elasticsearch.getContainer(),
+                    Arrays.asList(
+                            TestUtils.getResource("elasticsearch6-end-to-end-test.jar")
+                                    .toAbsolutePath()
+                                    .toUri()
+                                    .toURL(),
+                            TestUtils.getResource("flink-connector-testing-elasticsearch.jar")
+                                    .toAbsolutePath()
+                                    .toUri()
+                                    .toURL(),
+                            TestUtils.getResource("flink-connector-testing.jar")
+                                    .toAbsolutePath()
+                                    .toUri()
```
Can you explain why you need all three jars?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkExternalContext.java:
```diff
@@ -0,0 +1,135 @@
+
+    Elasticsearch6SinkExternalContext(
+            String addressExternal, String addressInternal, List<URL> connectorJarPaths) {
+        this.addressInternal = addressInternal;
+        this.connectorJarPaths = connectorJarPaths;
```
`checkNotNull`

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkExternalContext.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public ExternalSystemDataReader<TupleC2<String, String>> createSinkDataReader(
+            TestingSinkSettings sinkSettings) {
+        Elasticsearch6Utils.refreshIndex(client, indexName);
```
Would it make sense to move the refresh method into `Elasticsearch6DataReader#poll` ?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkExternalContext.java:
```diff
@@ -0,0 +1,135 @@
+                random.nextInt(NUM_RECORDS_UPPER_BOUND - NUM_RECORDS_LOWER_BOUND)
+                        + NUM_RECORDS_LOWER_BOUND;
+
+        for (int i = 0; i < recordNum; i++) {
+            int valueLength = random.nextInt(RANDOM_STRING_MAX_LENGTH) + 1;
+            String key = Integer.toString(i);
+            String value = RandomStringUtils.random(valueLength, true, true);
+            randomStringRecords.add(TupleC2.of(key, value));
+        }
+        return randomStringRecords;
```
Nit: Again, please use the stream syntax.

afedulov replied:
Copy-paste from Kafka tests. Migrated.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkExternalContext.java:
```diff
@@ -0,0 +1,135 @@
+            LOG.error("Cannot delete index {}", indexName, e);
+        }
+        // This is needed to avoid race conditions between tests that reuse the same index
+        Elasticsearch6Utils.refreshIndex(client, indexName);
```
Out of curiosity since every time you instantiate the DataReader the indices are refreshed can you give some more details about the race condition?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6SinkExternalContextFactory.java:
```diff
@@ -0,0 +1,51 @@
+
+    Elasticsearch6SinkExternalContextFactory(
+            ElasticsearchContainer elasticsearchContainer, List<URL> connectorJars) {
+        this.elasticsearchContainer = elasticsearchContainer;
+        this.connectorJars = connectorJars;
```
`checkNotNull`

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/com/apache/flink/streaming/tests/Elasticsearch6Utils.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    static void flushIndex(RestHighLevelClient client, String indexName) {
```
This method seems to be unused.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch7/pom.xml:
```diff
@@ -0,0 +1,158 @@
+			<version>1.16-SNAPSHOT</version>
+			<scope>test</scope>
+		</dependency>
+		<!--		<dependency>
```
Why does this module not depend on elasticsearch common?

afedulov replied:
I restructured the modules structure in the new version and cleaned up the poms.

fapaul reviewed (commented):
It is important to reduce the code duplication between elasticsearch6 and elasticsearch7. Having two dedicated test classes is a good idea, but you can share almost all utilities and only have thin wrappers around the different APIs.

afedulov commented:
Thanks a lot for your review @fapaul. I originally did not want to invest much time into addressing code duplication because I was under the impression that the ES6-related tests are going to be removed soon after the connector externalization. However I still see your point in doing so. At least when ES8 will be introduced, we'll be able to make use of the proper setup.
I reworked the structure and extracted everything up until the level when a lot of concrete Elasticsearch classes are used. Further deduplication between the `Elasticsearch6Client` and `Elasticsearch7Client` is potentially possible but would require reintroducing a facade for a huge chunk of low-level ES abstractions and it does not seem to be worth it.

afedulov commented:
@flinkbot run azure

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch-common/src/main/java/org/apache/flink/streaming/tests/ElasticsearchDataReader.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public List<KeyValue<Integer, String>> poll(Duration timeout) {
+        client.refreshIndex(indexName);
+        // TODO: Tests are flaky without this small delay.
```
I thought we had fixed this problem. Can you explain what is going on here?

afedulov replied:
https://github.com/apache/flink/pull/19405#discussion_r848945151

fapaul replied:
You need to fix this because the current timeout can easily become flaky in the future did you try setting the refresh policy [1] in the emitters that every write needs to wait until the index is refreshed? 

[1] https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-high-document-index.html#_optional_arguments

afedulov replied:
I will try, but I suspect that the issue manifests itself between the time when the writes are done from the Flink side and the test immediately starts the reader, which somehow still does not see the written data, even despite an active blocking index refresh. As the policy is related to writes, not reads, I do not expect it to change much.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/pom.xml:
```diff
@@ -0,0 +1,127 @@
+				<executions>
+					<execution>
+						<goals>
+							<goal>shade</goal>
```
Quick reminder why do we need shading here?

afedulov replied:
This is the equivalent of building a fat jar for the user program. If we where to use hte `flink-connector-elasticsearchX` dependency here instead https://github.com/apache/flink/blob/4015a3a7ae3f6cd6d05def3ac06f8b8dccbf31de/flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/test/java/org/apache/flink/streaming/tests/Elasticsearch6SinkE2ECase.java#L64-L71
, we'd also have to add every library that ES connector depends on as a separate JAR (and there are a LOT of dependencies).

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/main/java/org/apache/flink/streaming/tests/Elasticsearch6Client.java:
```diff
@@ -0,0 +1,123 @@
+                                            searchHit.getSourceAsMap().get("value").toString()))
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            return Collections.emptyList();
```
Add a log statement to increase the awareness of the error.

afedulov replied:
It is not really an exception, the sink test framework is designed to pull data until it is there.
https://github.com/apache/flink/blob/91c2b8d56bae03c2ce4a50b5c014cea842df9f74/flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SinkTestSuiteBase.java#L460

This is how it is also handled in the Kafka Sink tests:
https://github.com/apache/flink/blob/57e3f03ccd719ed772c983ba335517d95f8f3e6a/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/testutils/KafkaDataReader.java#L54

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch7/src/main/java/org/apache/flink/streaming/tests/Elasticsearch7Client.java:
```diff
@@ -0,0 +1,123 @@
+                                            searchHit.getSourceAsMap().get("value").toString()))
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            return Collections.emptyList();
```
Again add log

afedulov replied:
Same as https://github.com/apache/flink/pull/19405#discussion_r853967062

fapaul commented on flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SinkTestSuiteBase.java:
```diff
@@ -508,6 +509,7 @@ private void checkResultWithSemantic(
                                 .matchesRecordsFromSource(Arrays.asList(sort(testData)), semantic);
                         return true;
                     } catch (Throwable t) {
+                        LOG.error("Ooops", t);
```
Please use a more descriptive error message.

afedulov replied:
Whoops, this was not supposed to be merged. The way the test is designed the exception is swallowed by default, this is a remainder of my debugging.

afedulov replied:
Please notice that without this log, the tests will simply accidentally hang if you were to run them without "sleep" in the reader.

fapaul commented on flink-end-to-end-tests/run-nightly-tests.sh:
```diff
@@ -199,9 +199,6 @@ function run_group_2 {
 
     run_test "Netty shuffle direct memory consumption end-to-end test" "$END_TO_END_DIR/test-scripts/test_netty_shuffle_memory_control.sh"
 
-    run_test "Elasticsearch (v6.8.20) sink end-to-end test" "$END_TO_END_DIR/test-scripts/test_streaming_elasticsearch.sh 6 https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-6.8.20.tar.gz"
```
It would be great to remove the bash e2e tests on a separate commit.

fapaul reviewed (commented):
I think we should be consistent in naming the common modules. The Kafka module is called `[...]common-kafka` and yours is called  `[...]elasticsearch-common`.

I also left some more inline comments and please take a look at the CI that is failing at the moment. PTAL

afedulov commented on flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SinkTestSuiteBase.java:
```diff
@@ -500,6 +500,7 @@ private void checkResultWithSemantic(
             ExternalSystemDataReader<T> reader, List<T> testData, CheckpointingMode semantic)
             throws Exception {
         final ArrayList<T> result = new ArrayList<>();
+        int failed = 0;
```
Leftover from debugging. Removed.

afedulov commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch7/pom.xml:
```diff
@@ -0,0 +1,128 @@
+					<execution>
+						<phase>package</phase>
+						<goals>
+							<goal>shade</goal>
```
Same story https://github.com/apache/flink/pull/19405#discussion_r853980008

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch-common/src/main/java/org/apache/flink/streaming/tests/KeyValue.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public int compareTo(KeyValue<K, V> other) {
```
Ah, I see thanks for the hint. WDYT about adding the `compareTo` method to `Tuple2`?

afedulov replied:
That could work but would make `Tuple2` somewhat special, not sure if we want this for consistency reasons. An alternative is to define `Tuple` as comparable, but that would mean adding `compareTo` to 73 classes, including test utilities across the framework. Having a separate `KeyValue` class seems like a good alternative to both of those approaches. WDYT?

fapaul replied:
Let's keep like it currently is.

fapaul commented:
Are the `Elasticsearch6SinkExample` and `Elasticsearch7SinkExample` still used after removing the bash tests? AFAIK the test suites do not use it.

fapaul commented:
> I think we should be consistent in naming the common modules. The Kafka module is called `[...]common-kafka` and yours is called `[...]elasticsearch-common`.

Did you address this?

afedulov commented:
> Did you address this?
Pardon, missed the top review comment.  Fixed now.

afedulov commented:
> Are the Elasticsearch6SinkExample and Elasticsearch7SinkExample still used after removing the bash tests? AFAIK the test suites do not use it.

Good point. Cleaned up.

afedulov commented:
@flinkbot run azure

afedulov commented:
@flinkbot run azure

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/pom.xml:
```diff
@@ -43,50 +43,52 @@ under the License.
 						<phase>package</phase>
 						<goals>
-							<goal>shade</goal>
+							<goal>test-jar</goal>
```
Can you avoid introducing more test jars and maybe move the common utilities to the main package?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch7/pom.xml:
```diff
@@ -30,10 +30,14 @@ under the License.
-	<artifactId>flink-elasticsearch7-test</artifactId>
-	<name>Flink : E2E Tests : Elasticsearch 7</name>
+	<artifactId>flink-end-to-end-tests-elasticsearch7</artifactId>
+	<name>Flink : E2E Tests : Elasticsearch 7 Java</name>
```
Is `Java` really important here?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/src/test/java/org/apache/flink/streaming/tests/ElasticsearchSinkE2ECaseBase.java:
```diff
@@ -0,0 +1,92 @@
+                    .build();
+
+    @Override
+    protected void checkResultWithSemantic(
```
Can you explain this in more detail? Isn't this only checking that any record was written to elasticsearch?

fapaul commented on flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SinkTestSuiteBase.java:
```diff
@@ -540,7 +540,7 @@ private boolean compareSinkMetrics(
 
     /** Sort the list. */
-    private List<T> sort(List<T> list) {
+    protected List<T> sort(List<T> list) {
```
Not sure it makes much sense to make this method visible.

fapaul reviewed (commented):
Looks mostly good I left some last questions

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/pom.xml:
```diff
@@ -0,0 +1,82 @@
+	<dependencyManagement>
+	   <dependencies>
+		  <dependency>
+			 <groupId>org.apache.httpcomponents</groupId>
```
Can you add a small comment why this is necessary?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/src/main/java/org/apache/flink/streaming/tests/ElasticsearchSinkExternalContextBase.java:
```diff
@@ -0,0 +1,125 @@
+    public List<KeyValue<Integer, String>> generateTestData(
+            TestingSinkSettings sinkSettings, long seed) {
+        Random random = new Random(seed);
+        List<KeyValue<Integer, String>> randomStringRecords = new ArrayList<>();
```
Unused?

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/src/main/java/org/apache/flink/streaming/tests/ElasticsearchTestEmitter.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    default void emit(
+            KeyValue<Integer, String> element, SinkWriter.Context context, RequestIndexer indexer) {
+        addUpsertRequest(indexer, element);
```
I think it would be good to revisit the current `TestEmitter` structure. You can replace the inheritance with composition by only implementing one class that has only a constructor parameter to build the `UpdateRequest`. Currently, `emit` is marked as default, although all implementors override it, and sharing createDoc via a default method also feels strange because it could be a static method.

afedulov replied:
Fixed `emit` and `createDoc`, but I do not quite get which composition structure you propose.

fapaul replied:
Maybe I missed your change but `emit` looks similar to before. My idea is basically to only have one `ElasticsearchTestEmitter` implementing the emitter interface, and this class is used by elasticsearch6 and elasticsearch7. Both modules pass the needed information to construct the request in the constructor of the `ElasticsearchTestEmitter`. There also are still conflicts with the master branch.

afedulov replied:
Thanks for the clarification. Please let me know if the changes from the topmost commit are aligned with your idea.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch6/src/main/java/org/apache/flink/streaming/tests/Elasticsearch6Client.java:
```diff
@@ -0,0 +1,149 @@
+                LOG.info("Index already exists {}", indexName);
+            }
+        } catch (IOException e) {
+            LOG.error("Cannot create index {}", indexName);
```
Please always also include the exception when logging. This affects some places in this class.

fapaul commented on flink-end-to-end-tests/flink-end-to-end-tests-elasticsearch7/src/main/java/org/apache/flink/streaming/tests/Elasticsearch7Client.java:
```diff
@@ -0,0 +1,147 @@
+        try {
+            restClient.indices().refresh(refresh, RequestOptions.DEFAULT);
+        } catch (IOException e) {
+            LOG.error("Cannot delete index {}", indexName);
```
Same as for the other client, please include the exception in the log message.

fapaul reviewed (commented):
I left some minor comments. Only the TestEmitter structure might need another iteration, but everything else looks good. Please also resolve the conflicts with the master.

afedulov commented:
Resolved the conflicts.

fapaul reviewed (approved):
LGTM, can you squash the commits together?

afedulov commented:
Thanks, @fapaul, I squashed the review commits.
